### PR TITLE
feature: wasmModule and fsBundle options and caching of module+fsBundle

### PR DESCRIPTION
**Pull request by samwillis**
This adds additional options for the wasmModule and fsBundle, along with caching of the default ones if used.

This should enable a route to running in a Cloudflare worker (pending further changes) as you can do something like:

```ts
import { PGlite } from '@electric-sql/pglite';
import wasm from '@electric-sql/pglite/dist/postgres.wasm';
import fsBundle from '@electric-sql/pglite/dist/postgres.wasm';

pg = await PGlite.create({
	wasmModule: wasm,
	fsBundle: fsBinary,
});
```

We had something working similar to this in the old v0.1.n version of PGlite on Cloudflare

Fixes: #329 #81

### CODE REVIEW

File: packages/pglite/src/pglite.ts [hunk 1]
```diff
@@ -293,6 +326,9 @@ export class PGlite
     }
     emscriptenOpts['pg_extensions'] = extensionBundlePromises
 
+    // Await the fs bundle
+    await fsBundleBufferPromise
```
Comment (msfstef) [inline packages/pglite/src/pglite.ts:330]: the only `await` between this and the initalization of this promise is the `this.fs!.emscriptenOpts` call - which also receives `emscriptenOpts` which contains a reference to the `fsBundleBuffer` which might or might not have resolved (?)

I'm assuming you're familiar with when `getPreloadedPackage` is called and this is why it's safe, otherwise I would imagine that it's not too expensive to just wait for the FS bundle where we initialize the promise for simplicity?

File: packages/pglite/src/utils.ts [hunk 2]
```diff
@@ -5,30 +5,74 @@ export const IN_NODE =
...
-        url = PGSHARE_URL
-        break
-      case 'postgres.wasm':
-        url = PGWASM_URL
-        break
-      default:
-        console.error('makeLocateFile', base)
-    }
+  const moduleUrl = new URL('../release/postgres.wasm', import.meta.url)
+  wasmDownloadPromise = fetch(moduleUrl)
```
Comment (msfstef) [inline packages/pglite/src/utils.ts:15]: nit: guard against multiple calls here?

(also, I'm trembling at potential bugs here haha but I don't see why this wouldn't work - did CORS or other headers get added to bundle requests before?)

**Review by msfstef — APPROVED** [review]
Nice! Left a couple of questions - the starting of a promise and awaiting it later on while using the results of it within an intermediate stage is a bit hard to follow and I suspect hard to debug if something does go wrong - might be worth adding a comment or two where the FS is used in emscripten opts?

File: packages/pglite/src/pglite.ts (hunk 1 above)
Comment (samwillis) [reply packages/pglite/src/pglite.ts:330]: Yep, it's a little confusing. `getPreloadedPackage` is called when we do:

```ts
this.mod = await PostgresModFactory(emscriptenOpts)
```

The intention is too try and download the WASM, the FS bundle and extensions all in parallel. I'll add some comments.

File: packages/pglite/src/utils.ts (hunk 2 above)
Comment (samwillis) [reply packages/pglite/src/utils.ts:15]: :man_facepalming: I meant to add that....

A thought, should we also add a `fetchClient` similar to what we do in the Electric `ShapeStream` so that people can configure CORS headers?

Comment (msfstef) [reply packages/pglite/src/utils.ts:15]: eh, it feels like it errs on the side of overcustomization but I suppose this PR does address very particular resource acquisition cases, so might be a good idea?
